packages/styled-base/types/tests.tsx

```diff
@@ -217,6 +214,7 @@ const StyledClass1 = StyledClass0.withComponent(ReactClassComponent1)
 declare const ref1_0: (element: ReactClassComponent1 | null) => void
 declare const ref1_1: (element: ReactClassComponent0 | null) => void
 declare const ref1_2: (element: HTMLDivElement | null) => void
+// $ExpectError
```

Comment:
This is now an error because column doesn't exist on the new component, only the original

Reply:
I'm wondering if this is a correct change - while that prop is not existing on the new component the created interpolations are being carried over to the new component and that prop could be used in previous interpolations, right? In such a situation this could fail miserably at runtime, right?

In general withComponent & as are really difficult to type correctly due to their dynamic(-ish) semantics.

Reply:
Oh, right. I get that, I have to build up an intersection of the previous component props.

Nice catch!

Reply:
Fixed this now.

I have restructured the generic type params in StyledComponent and related types to make it really clear and the change allows ComponentProps from a normal component to get transferred while dropping tag props